### PR TITLE
VFS-609: VFS SFTP doesn't support a private key as byte array

### PR DESCRIPTION
**Pull request by RoTkay**
I really need feature with specifying private key in bytes. I guess that I'm not alone with this need.
Please, check my pull request. It's my first time making pull request in this project. If I made something wrong, I would be glad to hear your comments.
P.S. I took code changes from https://issues.apache.org/jira/browse/VFS-609 that stevezhuang attached to ticket, but that code wasn't placed to repo.

### CODE REVIEW

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 1]
```diff
@@ -69,6 +71,43 @@ public IdentityInfo(final File privateKey, final File publicKey, final byte[] pa
...
         this.publicKey = publicKey;
         this.passPhrase = passPhrase;
+        this.privateKeyBytes = null;
+        this.publicKeyBytes = null;
+    }
+
+
+    /**
+     * Constructs an identity info with private and public key and passphrase for the private key.
+     * <p>
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: Open para tag is not needed.

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 2]
```diff
@@ -69,6 +71,43 @@ public IdentityInfo(final File privateKey, final File publicKey, final byte[] pa
...
+     *
+     * @param privateKey Private key bytes
+     * @param publicKey The public key part used for connections with exchange of certificates (can be {@code null})
+     * @param passPhrase The passphrase to decrypt the private key (can be {@code null} if no passphrase is used)
+     * @since 2.1
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: The next version will be 2.4, not 2.1.

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 3]
```diff
@@ -69,6 +71,43 @@ public IdentityInfo(final File privateKey, final File publicKey, final byte[] pa
...
+        this.privateKey = null;
+        this.publicKey = null;
+        this.privateKeyBytes = privateKey;
+        this.publicKeyBytes = publicKey;
+        this.passPhrase = passPhrase;
+    }
+
+    /**
+     * Constructs an identity info with private and passphrase for the private key.
+     * <p>
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: See above.

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 4]
```diff
@@ -69,6 +71,43 @@ public IdentityInfo(final File privateKey, final File publicKey, final byte[] pa
...
+     *
+     * @param privateKey Private key bytes
+     * @param passPhrase The passphrase to decrypt the private key (can be {@code null} if no passphrase is used)
+     * @since 2.1
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: See above.

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 5]
```diff
@@ -100,4 +139,24 @@ public File getPublicKey() {
     public byte[] getPassPhrase() {
         return passPhrase;
     }
+
+    /**
+     * Get private key bytes.
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: Javadoc should use "Gets..."

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java [hunk 6]
```diff
@@ -100,4 +139,24 @@ public File getPublicKey() {
...
+     *
+     * @return private key
+     */
+    public byte[] getPrivateKeyBytes()
+    {
+        return privateKeyBytes;
+    }
+
+    /**
+     * Get public key bytes.
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/IdentityInfo.java]: See above.

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java [hunk 7]
```diff
@@ -195,11 +195,17 @@ private static void setConfigRepository(final JSch jsch, final File sshDir, fina
         }
     }
 
-    private static void addIndentity(final JSch jsch, final IdentityInfo info) throws FileSystemException {
+    private static void addIdentity(final JSch jsch, final IdentityInfo info) throws FileSystemException {
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java]: Good catch! :-)

File: commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java [hunk 8]
```diff
@@ -195,11 +195,17 @@ private static void setConfigRepository(final JSch jsch, final File sshDir, fina
         }
     }
 
-    private static void addIndentity(final JSch jsch, final IdentityInfo info) throws FileSystemException {
+    private static void addIdentity(final JSch jsch, final IdentityInfo info) throws FileSystemException {
         try {
-            final String privateKeyFile = info.getPrivateKey() != null ? info.getPrivateKey().getAbsolutePath() : null;
-            final String publicKeyFile = info.getPublicKey() != null ? info.getPublicKey().getAbsolutePath() : null;
-            jsch.addIdentity(privateKeyFile, publicKeyFile, info.getPassPhrase());
+            if (info.getPrivateKey() != null) {
```
Comment (garydgregory) [inline commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java]: It is confusing if more than one setting is used but one overrides the other silently. Wouldn't throwing an IllegalStateException be clearer?

Comment (RoTkay) [reply commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java]: Sorry, I didn't catch what do you mean by that. Throw` IllegalStateException` in which case?

Comment (garydgregory) [reply commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java]: What should happen if the `info` contains BOTH a private key file and private key bytes? Right now the file quietly overrides the bytes. This is good for compatibility. If neither are set you get an exception, which is new behavior. 
Should we throw an exception if both are set?
Is it possible for the provider to be used _without_ a PK?

Comment (garydgregory) [reply commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/sftp/SftpClientFactory.java]: Ah, I see. The issue I have now is that you have to know how which ctor you called to know which getters are valid, which is semi-lame. Let me see if I can adapt your code...

**Comment by garydgregory** [comment]
Please see git master and the new class `org.apache.commons.vfs2.provider.sftp.BytesIdentityInfo`.

**Comment by RoTkay** [comment]
Okey, saw your solution. Thanks for adding bytes approach! And thanks for your comments too :)

**Pull request closed without merging**